Inspect class to determine if its a class, actor, or distributed actor

heckj opened the pull request:
For the purposes of symbol graph generation

Updates the function printAbridgedType to inspect the declaration to determine if its a class, actor, or distributed actor in order to present "actor" or "distributed actor" into the symbol graph so that DocC (documentation) can represent the type correctly.

resolves https://github.com/apple/swift-docc/issues/212

heckj committed:
inspects class to determine if its a class, actor, or distributed actor for symbol graph generation

heckj commented:
@swift-ci Please test

QuietMisdreavus commented on lib/SymbolGraphGen/DeclarationFragmentPrinter.cpp:
```diff
@@ -189,9 +189,17 @@ void DeclarationFragmentPrinter::printAbridgedType(const GenericTypeDecl *TD,
-        break;
+        case DeclKind::Class: {
+            auto *newthing = cast<ClassDecl>(TD);
+            if (newthing->isActor()) {
+                printText("actor");
+            } else if (newthing->isDistributedActor()) {
+                printText("distributed actor");
```
Are we sure that distributed actors won't fall into the first branch? I'm unfamiliar with the underlying decl types so i dunno if this will cause problems.

heckj replied:
I'm not sure, but I'll definitely do my best to add tests for both an "actor" and "distributed actor", as well as re-read through the declaration types with the detail that Slava provided in the forums.

CodaFi replied:
These are semantic requests here - are you sure you want to go all the way through to protocol resolution to determine if something is declared as a `class` or an `actor` or a `distributed actor`? Since you know it's a `ClassDecl`, you can use `newthing->isExplicitActor()` and `newthing->isExplicitDistributedActor()` to do a syntactic check.

CodaFi replied:
The semantic request `isActor()` will, for example, see this as an actor

```
class MyCustomActor: Actor { /**/ }
```

and this as a distributed actor

```
class MyDistributedActor: DistributedActor { /**/ }
```

heckj replied:
Ah, thank you @CodaFi - I'll switch that up, sounds like I touched on the wrong API here for what I was after.

QuietMisdreavus reviewed (commented):
Are you able to write a test for this? You could probably get away with copying from [`test/SymbolGraph/Symbols/Mixins/DeclarationFragments/Full/NominalTypes.swift`](https://github.com/apple/swift/blob/main/test/SymbolGraph/Symbols/Mixins/DeclarationFragments/Full/NominalTypes.swift) and writing basic `actor` and `distributed actor` declarations so you can double-check that it still prints correctly. I can help you out on the Swift Slack if you need pointers for building and running the tests.

QuietMisdreavus commented:
@swift-ci Please smoke test

Kyle-Ye commented on lib/SymbolGraphGen/DeclarationFragmentPrinter.cpp:
```diff
@@ -189,9 +189,17 @@ void DeclarationFragmentPrinter::printAbridgedType(const GenericTypeDecl *TD,
+            if (newthing->isActor()) {
+                printText("actor");
+            } else if (newthing->isDistributedActor()) {
+                printText("distributed actor");
```
+1 for adding `isExplicitActor` method here to force the semantic instead of depending the ordering.

Kyle-Ye commented on lib/SymbolGraphGen/DeclarationFragmentPrinter.cpp:
```diff
@@ -189,9 +189,17 @@ void DeclarationFragmentPrinter::printAbridgedType(const GenericTypeDecl *TD,
+        case DeclKind::Class: {
+            auto *newthing = cast<ClassDecl>(TD);
+            if (newthing->isActor()) {
+                printText("actor");
```
I do not know if `kw_action` exists. But if not, I suggest we should add it first and use the constant instead of using a hard cord string here.

```suggestion
                printText(tok::kw_actor);
```

QuietMisdreavus replied:
It looks like [`TokenKinds.def`](https://github.com/apple/swift/blob/main/include/swift/AST/TokenKinds.def) doesn't define an `actor` keyword, so unfortunately we'll need to spell it out. 😞

QuietMisdreavus replied:
That definition would go in [`TokenKinds.def`](https://github.com/apple/swift/blob/main/include/swift/AST/TokenKinds.def), but i would want to make extra sure that we can add the keyword there before doing it. I'm not sure who to ask, though.

heckj replied:
I've added a keyword declaration in for "actor", but I'm hesitant to do so:
1 - The token wasn't added earlier, but clearly the parser understands the keyword `actor` already, and it's not clear why it wasn't added back when `actor` first became a thing.
2 - I don't seem to be able to do the same with 'distributed', the modifier for 'actor', which makes the remaining piece asserting distributed actor or not sort of awkward.
3 - I'm uncertain of the implications of adding a token to the keyword list, and where else that ripples into. The parser at least. Looking through the parser, "actor" and "distributed actor" are handled are raw strings, not using `Tok::` structures.

A full build locally succeeded with the `tok::actor` added in there, as did the lit.py tests. (well, 3 failed, but the same three as the first run at this, and seemingly unrelated):

```
Failed Tests (3):
  Swift(macosx-arm64) :: Reflection/capture_descriptors.sil
  Swift(macosx-arm64) :: stdlib/CharacterRecognizer.swift
  Swift(macosx-arm64) :: stdlib/StringIndex.swift
```

heckj replied:
I added it in, just to try - but I can very easily pull it back out, and as I mentioned, I'm hesitant to go here because it hasn't been added previously, with the parser seeming to use the older keyword sets, but not exclusively.

heckj replied:
After creating the test and trying to verify it, I was having all sorts of trouble compiling everything, so I backed out that specific change and reverted to printing the raw string rather than using a token.

ktoso replied:
Hmm, I don't know about the implications of adding the keyword there, maybe @ahoppen could advise?

slavapestov replied:
The parser handles `actor` and `distributed` as strings and not keywords, because they're not keywords. For example, the following is valid:

```
struct actor {}
var x: actor = actor()
```

ahoppen replied:
`actor` and `distributed` are contextual keywords and thus shouldn’t be in TokenKinds.def. Notice how `await`, `async` and all sorts of other keywords that were introduced after Swift 2 or 3 also aren’t in TokenKinds.def.

ktoso replied:
Ah I see, thanks for chiming in! :-)

heckj replied:
Thank you!

Kyle-Ye commented on lib/SymbolGraphGen/DeclarationFragmentPrinter.cpp:
```diff
@@ -189,9 +189,17 @@ void DeclarationFragmentPrinter::printAbridgedType(const GenericTypeDecl *TD,
-      case DeclKind::Class:
-        printText(getTokenText(tok::kw_class));
-        break;
+        case DeclKind::Class: {
```
The indent is different of the above statements. Is is on purpose or a mistake?

heckj replied:
Heh - Xcode doesn't always maintain formatting well. This should be fixed in an update shortly.

Kyle-Ye replied:
Reopen the conversation as the indent still looks wrong to me.

heckj replied:
Huh - thanks @Kyle-Ye, I'll look closer at tabs vs spaces here, I thought I'd fixed it, but I'm new to contributing here.

Kyle-Ye replied:
If you are using Xcode as your IDE, you may change the setting(Tab Width and Indent Width) here to match the existing code.

<img width="942" alt="image" src="https://github.com/apple/swift/assets/43724855/7442db7a-3184-4f46-b8b9-d9dd93df348a">

ktoso commented on lib/SymbolGraphGen/DeclarationFragmentPrinter.cpp:
```diff
@@ -189,9 +189,17 @@ void DeclarationFragmentPrinter::printAbridgedType(const GenericTypeDecl *TD,
+          auto *refDecl = cast<ClassDecl>(TD);
+          if (refDecl->isExplicitActor()) {
+            printText("actor");
+          } else if (refDecl->isExplicitDistributedActor()) {
```
yeah the `Explicit` are the right calls to make here, thank you!

ktoso commented:
Thanks for looking into this one, I've long forgotten about it huh!

You'll want to include a test for the distributed actor type as well, something like this:

```
// RUN: %empty-directory(%t)
// RUN: %target-build-swift %s -module-name Actor -emit-module -emit-module-path %t/
// RUN: %target-swift-symbolgraph-extract -module-name Actor -I %t -pretty-print -output-dir %t
// RUN: %FileCheck %s --input-file %t/Actor.symbols.json

// REQUIRES: distributed
import Distributed 

// CHECK: "identifier": "swift.class"
// CHECK-NEXT: "displayName": "Class"
// CHECK: pathComponents
// CHECK-NEXT: "DA"
// CHECK: "kind": "keyword"
// CHECK: "spelling": "distributed actor"
public distributed actor DA {
  public typealias ActorSystem = LocalTestingDistributedActorSystem
}


```

heckj commented:
Both tests (Actor and DistributedActor) added, lit tests passing for me locally. I found I had to switch around the logic to go from more specific to less specific, but with that tweak its passing now. I've left out adding a token - if that's something we would prefer to add, I'm happy to do so, but since the existing parser **isn't** using a token, I thought it best to stick with raw strings - plus a tok::kw_* wouldn't help with `distributed actor`.

QuietMisdreavus reviewed (approved):
This looks great! Thank you so much for picking this up.

QuietMisdreavus commented:
@swift-ci Please smoke test

QuietMisdreavus commented:
I'm fine with not adding the keyword for now; maybe in an "all SwiftSyntax all the time" future we can revisit this, but for now this is a good fix for a long-standing issue.

heckj commented:
Added a follow-up commit to tack on explicit availability to resolve a warning from the smoke test

ktoso commented:
@swift-ci please smoke test

heckj commented:
@ktoso The windows SDK is failing while attempting to compile the distributed actor snippet:

```
command stderr:
<unknown>:0: error: cannot load underlying module for 'WinSDK'
Error: Failed to load the module 'DistributedActor'. Are you missing build dependencies or include/framework directories?
See the previous error messages for details. Aborting.
```

It worked fine for Actor, and I set up similar availability stanzas for the two types. Since you’ve worked more extensively with Distributed, and suggestions for a path forward? Do I need an addition optional import to load in supporting pieces or such?

ktoso commented:
@swift-ci please smoke test Windows

ktoso commented:
I'll help investigate the Windows failure, need to re-run though

heckj commented:
I captured the output from the failure of the DistributedActor.swift `lit` test in the PR - full text this time, and a broader selection of the output from the re-run:

```
FAIL: Swift(windows-x86_64) :: SymbolGraph/Symbols/Kinds/DistributedActor.swift (8838 of 10873)
******************** TEST 'Swift(windows-x86_64) :: SymbolGraph/Symbols/Kinds/DistributedActor.swift' FAILED ********************
Script:
--
: 'RUN: at line 1';   rm -rf "T:\5\tools\swift\test-windows-x86_64\SymbolGraph\Symbols\Kinds\Output\DistributedActor.swift.tmp" && mkdir -p "T:\5\tools\swift\test-windows-x86_64\SymbolGraph\Symbols\Kinds\Output\DistributedActor.swift.tmp"
: 'RUN: at line 2';   't:\\5\\bin\\swiftc.exe' -target x86_64-unknown-windows-msvc  -vfsoverlay T:/5/tools/swift\stdlib\windows-vfs-overlay.yaml -strict-implicit-module-context -Xcc -Xclang -Xcc -fbuiltin-headers-in-system-modules -swift-version 4  -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 9999:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.0:macOS 10.14.4, iOS 12.2, watchOS 5.2, tvOS 12.2' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.1:macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.2:macOS 10.15.4, iOS 13.4, watchOS 6.2, tvOS 13.4' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.3:macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.4:macOS 11.3, iOS 14.5, watchOS 7.4, tvOS 14.5' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.5:macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.6:macOS 12.3, iOS 15.4, watchOS 8.5, tvOS 15.4' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.7:macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.8:macOS 13.3, iOS 16.4, watchOS 9.4, tvOS 16.4' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.9:macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.10:macOS 14.4, iOS 17.4, watchOS 10.4, tvOS 17.4, visionOS 1.1' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 6.0:macOS 15.0, iOS 18.0, watchOS 11.0, tvOS 18.0, visionOS 2.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 6.1:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, visionOS 9999'  -libc MD C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\SymbolGraph\Symbols\Kinds\DistributedActor.swift -module-name DistributedActor -emit-module -emit-module-path T:\5\tools\swift\test-windows-x86_64\SymbolGraph\Symbols\Kinds\Output\DistributedActor.swift.tmp/
: 'RUN: at line 3';   't:\\5\\bin\\swift-symbolgraph-extract.exe' -target x86_64-unknown-windows-msvc -module-cache-path T:\5\swift-test-results\x86_64-unknown-windows-msvc\clang-module-cache -module-name DistributedActor -I T:\5\tools\swift\test-windows-x86_64\SymbolGraph\Symbols\Kinds\Output\DistributedActor.swift.tmp -pretty-print -output-dir T:\5\tools\swift\test-windows-x86_64\SymbolGraph\Symbols\Kinds\Output\DistributedActor.swift.tmp
: 'RUN: at line 4';   "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python39_64\python.exe" C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\PathSanitizingFileCheck --allow-unused-prefixes --sanitize BUILD_DIR=T:/5/tools/swift --sanitize SOURCE_DIR=C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/swift --use-filecheck t:\5\bin\filecheck.exe  --enable-windows-compatibility C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\SymbolGraph\Symbols\Kinds\DistributedActor.swift --input-file T:\5\tools\swift\test-windows-x86_64\SymbolGraph\Symbols\Kinds\Output\DistributedActor.swift.tmp/DistributedActor.symbols.json
--
Exit Code: 1

Command Output (stdout):
--
$ ":" "RUN: at line 1"
$ "rm" "-rf" "T:\5\tools\swift\test-windows-x86_64\SymbolGraph\Symbols\Kinds\Output\DistributedActor.swift.tmp"
$ "mkdir" "-p" "T:\5\tools\swift\test-windows-x86_64\SymbolGraph\Symbols\Kinds\Output\DistributedActor.swift.tmp"
$ ":" "RUN: at line 2"
$ "t:\\5\\bin\\swiftc.exe" "-target" "x86_64-unknown-windows-msvc" "-vfsoverlay" "T:/5/tools/swift\stdlib\windows-vfs-overlay.yaml" "-strict-implicit-module-context" "-Xcc" "-Xclang" "-Xcc" "-fbuiltin-headers-in-system-modules" "-swift-version" "4" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 9999:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 5.0:macOS 10.14.4, iOS 12.2, watchOS 5.2, tvOS 12.2" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 5.1:macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 5.2:macOS 10.15.4, iOS 13.4, watchOS 6.2, tvOS 13.4" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 5.3:macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 5.4:macOS 11.3, iOS 14.5, watchOS 7.4, tvOS 14.5" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 5.5:macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 5.6:macOS 12.3, iOS 15.4, watchOS 8.5, tvOS 15.4" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 5.7:macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 5.8:macOS 13.3, iOS 16.4, watchOS 9.4, tvOS 16.4" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 5.9:macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 5.10:macOS 14.4, iOS 17.4, watchOS 10.4, tvOS 17.4, visionOS 1.1" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 6.0:macOS 15.0, iOS 18.0, watchOS 11.0, tvOS 18.0, visionOS 2.0" "-Xfrontend" "-define-availability" "-Xfrontend" "SwiftStdlib 6.1:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, visionOS 9999" "-libc" "MD" "C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\SymbolGraph\Symbols\Kinds\DistributedActor.swift" "-module-name" "DistributedActor" "-emit-module" "-emit-module-path" "T:\5\tools\swift\test-windows-x86_64\SymbolGraph\Symbols\Kinds\Output\DistributedActor.swift.tmp/"
$ ":" "RUN: at line 3"
$ "t:\\5\\bin\\swift-symbolgraph-extract.exe" "-target" "x86_64-unknown-windows-msvc" "-module-cache-path" "T:\5\swift-test-results\x86_64-unknown-windows-msvc\clang-module-cache" "-module-name" "DistributedActor" "-I" "T:\5\tools\swift\test-windows-x86_64\SymbolGraph\Symbols\Kinds\Output\DistributedActor.swift.tmp" "-pretty-print" "-output-dir" "T:\5\tools\swift\test-windows-x86_64\SymbolGraph\Symbols\Kinds\Output\DistributedActor.swift.tmp"
# command stderr:
<unknown>:0: error: cannot load underlying module for 'WinSDK'
Error: Failed to load the module 'DistributedActor'. Are you missing build dependencies or include/framework directories?
See the previous error messages for details. Aborting.
```

Digging around on that specific error, I found a forums post from 2020 that talks about efforts needed at that time to [expose the windows SDK up to the relevant module](https://forums.swift.org/t/cannot-load-underlying-module-for-winsdk/40123/2), the main gist of that content (which I'm not sure is still relevant):

```
set SDKROOT=%SystemDrive%\Library\Developer\Platforms\Windows.platform\Developer\SDKs\Windows.sdk
copy "%SDKROOT%\usr\share\ucrt.modulemap" "%UniversalCRTSdkDir%\Include\%UCRTVersion%\ucrt\module.modulemap"
copy "%SDKROOT%\usr\share\visualc.modulemap" "%VCToolsInstallDir%\include\module.modulemap"
copy "%SDKROOT%\usr\share\visualc.apinotes" "%VCToolsInstallDir%\include\visualc.apinotes"
copy "%SDKROOT%\usr\share\winsdk.modulemap" "%UniversalCRTSdkDir%\Include\%UCRTVersion%\um\module.modulemap"
```

Since this is really exercising `swift-symbolgraph-extract.exe`, is there additional module cache path options, or something equivalent, that needs to be put in place to expose the WinSDK modulemap?

QuietMisdreavus commented:
@swift-ci Please smoke test Windows

QuietMisdreavus commented:
@swift-ci Please smoke test Linux

QuietMisdreavus commented:
@swift-ci Please smoke test macOS

QuietMisdreavus commented:
@swift-ci Please smoke test

QuietMisdreavus commented:
@swift-ci Please smoke test

ktoso commented:
@swift-ci please smoke test

QuietMisdreavus commented:
@swift-ci Please smoke test

ktoso commented:
Thanks a lot for figuring it out @heckj ! LGTM

heckj commented:
rebased down to a single commit, and added a final commit in the lit.cfg manipulation to explain why its there in case someone else runs down that road in the future. Should be ready now.

ktoso commented:
@swift-ci please smoke test

QuietMisdreavus commented:
@heckj Thanks again for working on this! I'm super glad we were able to get this in.
